### PR TITLE
[MERGE] Merge 1.7

### PR DESCRIPTION
**Pull request by gpang**
The conflicts are resolved in the last commit.

### CODE REVIEW

File: core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java [hunk 1]
```diff
@@ -133,7 +137,7 @@ public void before() throws Exception {
       when(mBlockStore.getEligibleWorkers())
           .thenReturn(Arrays.asList(new BlockWorkerInfo(new WorkerNetAddress(), 0, 0)));
       when(mBlockStore.getInStream(eq((long) i), any(InStreamOptions
-          .class)))
+          .class), any()))
```
Comment (gpang) [inline core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java]: @bf8086 can you check if the conflict resolution is correct?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 2]
```diff
@@ -180,6 +180,8 @@ public synchronized void connect() throws AlluxioStatusException {
       try {
         mAddress = getAddress();
       } catch (UnavailableException e) {
+        LOG.warn("Failed to determine master RPC address ({}), retrying: {}",
```
Comment (gpang) [inline core/common/src/main/java/alluxio/AbstractClient.java]: @aaudiber can you check if the conflict resolution is correct?

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java [hunk 3]
```diff
@@ -99,12 +110,23 @@ public int read() throws IOException {
     if (mPosition == mLength) { // at end of file
       return -1;
     }
-    updateStream();
-    int result = mBlockInStream.read();
-    if (result != -1) {
-      mPosition++;
+    CountingRetry retry = new CountingRetry(MAX_WORKERS_TO_RETRY);
+    IOException lastException = null;
+    while (retry.attempt()) {
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: @bf8086 Please check if this still correct.

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java [hunk 4]
```diff
@@ -126,13 +148,23 @@ public int read(byte[] b, int off, int len) throws IOException {
...
+      try {
+        int bytesRead = mBlockInStream.read(b, currentOffset, bytesLeft);
+        if (bytesRead > 0) {
+          bytesLeft -= bytesRead;
+          currentOffset += bytesRead;
+          mPosition += bytesRead;
+        }
+        retry.reset();
+      } catch (UnavailableException | ConnectException | DeadlineExceededException e) {
+        if (!retry.attempt()) {
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: @bf8086 Please check if this still correct.

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java [hunk 5]
```diff
@@ -172,21 +204,29 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
...
         long offset = pos % mBlockSize;
         int bytesRead =
             stream.positionedRead(offset, b, off, (int) Math.min(mBlockSize - offset, len));
         Preconditions.checkState(bytesRead > 0, "No data is read before EOF");
         pos += bytesRead;
         off += bytesRead;
         len -= bytesRead;
+        retry.reset();
+      } catch (UnavailableException | DeadlineExceededException | ConnectException e) {
+        if (!retry.attempt()) {
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: @bf8086 Please check if this still correct.

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 2 above)
Comment (aaudiber) [reply core/common/src/main/java/alluxio/AbstractClient.java]: LGTM

File: core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java (hunk 1 above)
Comment (bf8086) [reply core/client/fs/src/test/java/alluxio/client/file/FileInStreamTest.java]: LGTM.

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java (hunk 3 above)
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: LGTM.

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java (hunk 4 above)
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: This still works but the retry count semantic is changed. I will address it in a separate PR.

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java (hunk 5 above)
Comment (bf8086) [reply core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: same here.

**Pull request merged**